### PR TITLE
[WIP]Ferminet loss function and model prototype

### PR DESCRIPTION
**Pull request by shaipranesh2**
# Pull Request Template
Adds the loss function for ferminet, which is the hamiltonian function of the wavefunction(kinetic energy + potential energy)

## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by peastman** [comment]
Shouldn't it be doing all the calculations with PyTorch rather than Numpy?

**Comment by shaipranesh2** [comment]
@peastman @tonydavis629 Can I use the [functorch](https://github.com/pytorch/functorch) library's grad function, which allows us to pass it down to the torch's jacobian function to get the hessian.
 otherwise, we have to calculate the jacobian and hessian, by passing it into their functions separately(hence we calculate the jacobian twice!! and this step is like the rate determining step in the whole neural network)

**Review by tonydavis629 — APPROVED** [review]
@peastman @rbharath This PR looks ready to go

File: deepchem/models/torch_models/ferminet.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+from deepchem.utils.electron_sampler import ElectronSampler
+
+
+def test_f(x: np.ndarray) -> np.ndarray:
+  # dummy function which can be passed as the parameter f. f gives the log probability
+  # TODO replace this function with forward pass of the model in future
+  return 2 * np.log(np.random.uniform(low=0, high=1.0, size=np.shape(x)[0]))
+
+
+class FerminetModel(torch.nn.Module):
```
Comment (rbharath) [inline deepchem/models/torch_models/ferminet.py:30]: Doing a drive by review, so apologies if you've already discussed this. The usual convention we have is that classes ending in `Model` should inherit from `TorchModel`. So `FerminetModel` should inherit from `TorchModel`. Do you have a `TorchModel` wrapper in place yet?

Comment (shaipranesh2) [reply deepchem/models/torch_models/ferminet.py:30]: Thanks for the comment, I should interchange the 2 classes name here - Ferminet and FerminetModel

**Comment by shaipranesh2** [comment]
closing this PR due to merge conflicts b/w this and the ferminet input stream PR which got merged in.
